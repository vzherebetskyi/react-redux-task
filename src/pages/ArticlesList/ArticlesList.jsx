import React, { useState, useEffect } from 'react';
import { toast } from 'react-toastify';
import { useSelector, useDispatch } from 'react-redux';

import Layout from '../../components/layouts/base';
import './ArticlesList.scss';
import { getNewsAPINews, getGuardianAPINews, getNYTAPINews } from '../../services';
import { convertArticlesDataToUnifiedFormat } from '../../utils/helpers';
import { APITypes } from '../../utils/constants';
import { addArticles, purgeArticles } from '../../redux/actions/articlesActions';
import { updateFilters } from '../../redux/actions/filtersActions';
import ArticleCard from '../../components/ArticleCard/ArticleCard';
import SkeletonCard from '../../components/SkeletonCard/SkeletonCard';
import FilterBlock from '../../components/FilterBlock/FilterBlock';

const ArticlesList = () => {
  const articles = useSelector((state) => state.articlesReducer.filteredArticles);
  const searchValue = useSelector((state) => state.filtersReducer.searchValue);
  const dispatch = useDispatch();
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    getNews();
  }, []);

  useEffect(() => {
    dispatch(purgeArticles());
    getNews(searchValue);
  }, [searchValue]);

  const getNews = async (searchValue) => {
    setIsLoading(true);
    try {
      const newsArticles = await Promise.allSettled([
        getNewsAPINews(searchValue),
        getGuardianAPINews(searchValue),
        getNYTAPINews(searchValue),
      ]);
      const unifiedNewsApiArticles = convertArticlesDataToUnifiedFormat(
        newsArticles[0].value ? newsArticles[0].value.articles : [],
        APITypes[0],
      );
      const unifiedGuardianApiArticles = convertArticlesDataToUnifiedFormat(
        newsArticles[1].value ? newsArticles[1].value.articles : [],
        APITypes[1],
      );
      const unifiedNYTApiArticles = convertArticlesDataToUnifiedFormat(
        newsArticles[2].value ? newsArticles[2].value.articles : [],
        APITypes[2],
      );

      const unifiedFiltersData = {
        date: [
          ...new Set([
            ...unifiedNewsApiArticles[1].date.filter((el) => el),
            ...unifiedGuardianApiArticles[1].date.filter((el) => el),
            ...unifiedNYTApiArticles[1].date.filter((el) => el),
          ]),
        ],
        category: [
          ...new Set([
            ...unifiedNewsApiArticles[1].category.filter((el) => el),
            ...unifiedGuardianApiArticles[1].category.filter((el) => el),
            ...unifiedNYTApiArticles[1].category.filter((el) => el),
          ]),
        ],
        source: [
          ...new Set([
            ...unifiedNewsApiArticles[1].source.filter((el) => el),
            ...unifiedGuardianApiArticles[1].source.filter((el) => el),
            ...unifiedNYTApiArticles[1].source.filter((el) => el),
          ]),
        ],
      };

      dispatch(
        addArticles([...unifiedNewsApiArticles[0], ...unifiedGuardianApiArticles[0], ...unifiedNYTApiArticles[0]]),
      );
      dispatch(updateFilters(unifiedFiltersData));
    } catch (e) {
      toast.error(e.error);
    }
    setIsLoading(false);
  };

  return (
    <Layout>
      <FilterBlock />
      <div className="articles-wrapper">
        {isLoading ? (
          [...Array(5).keys()].map((el, idx) => <SkeletonCard key={`skeleton-${idx}`} />)
        ) : articles.length > 0 ? (
          articles.map((article, idx) => <ArticleCard key={`${article.api}-${idx}`} article={article} />)
        ) : (
          <p className="no-articles-message">No articles found</p>
        )}
      </div>
    </Layout>
  );
};

export default ArticlesList;
