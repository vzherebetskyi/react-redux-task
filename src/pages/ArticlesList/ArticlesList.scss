.articles-wrapper {
  display: flex;
  flex-direction: column;
  max-width: 100rem;
  margin: 0 auto;

  @media screen and (max-width: 1000px) {
    padding: 0 10px;
  }
}

.no-articles-message {
  margin: 1rem auto;
  font-size: 1.5rem;

  @media screen and (max-width: 1000px) {
    font-size: 1rem;
  }
}
