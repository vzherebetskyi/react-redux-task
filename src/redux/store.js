import { configureStore } from "@reduxjs/toolkit";

import rootReducer from "./reducers";
import rootState from "./rootState";

export default configureStore({reducer: rootReducer, preloadedState: rootState});
