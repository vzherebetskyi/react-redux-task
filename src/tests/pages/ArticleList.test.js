import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';

import store from '../../redux/store';
import ArticlesList from '../../pages/ArticlesList/ArticlesList';

test('should render ArticlesList component correctly', () => {
  const elem = renderer
    .create(
      <Provider store={store}>
        <ArticlesList />
      </Provider>,
    )
    .toJSON();
  expect(elem).toMatchSnapshot();
});
