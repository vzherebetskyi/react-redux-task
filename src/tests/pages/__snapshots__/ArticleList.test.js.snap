// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ArticlesList component correctly 1`] = `
<div>
  <div
    className="filter-wrapper "
  />
  <div
    className="articles-wrapper"
  >
    <p
      className="no-articles-message"
    >
      No articles found
    </p>
  </div>
  <div
    className="Toastify"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  />
</div>
`;
